File: julia/src/userfunc_mandelbrot.jl
using BenchmarkTools
using Test

include("utils.jl")

myabs2(z) = real(z) * real(z) + imag(z) * imag(z)

function mandel(z)
  maxiter = 80
  c = z
  for n in 0:(maxiter-1)
    if myabs2(z) > 4
      return n
    end
    z = z^2 + c
  end
  return maxiter
end

userfunc_mandelbrot() = [mandel(complex(r, i)) for i in -1.0:0.1:1.0, r in -2.0:0.1:0.5]

# Test output
@test sum(userfunc_mandelbrot()) == 14791

# Run benchmark
write_benchmark_result("userfunc_mandelbrot", @benchmark userfunc_mandelbrot())

File: benchmarks/python/micro_benchmarks/userfunc_mandelbrot.py
from micro_benchmarks.utils import benchmark, write_benchmark_result


def abs2(z: complex):
    return z.real * z.real + z.imag * z.imag


def mandel(z: complex) -> int:
    maxiter = 80
    c = z
    for n in range(maxiter):
        if abs2(z) > 4:
            return n
        z = z * z + c
    return maxiter


def userfunc_mandelbrot():
    r1 = [-2.0 + 0.1 * i for i in range(26)]
    r2 = [-1.0 + 0.1 * i for i in range(21)]
    return [mandel(complex(r, i)) for r in r1 for i in r2]


# Test output
assert sum(userfunc_mandelbrot()) == 14791

# Run benchmark
write_benchmark_result(
    "userfunc_mandelbrot", benchmark(userfunc_mandelbrot, n_trials=1000, n_warmups=500)
)
